### ISSUE TITLE
Problem with barrier / HCOLL

### ISSUE THREAD

**Issue by angainor** [open-mpi/ompi#6917]
I see a problem with HCOLL and `MPI_ Barrier` on an IB system running HPCX 2.4. I tested openmpi 4.0.1 and 3.1.3, and see the same behavior: when one of the ranks sleeps before entering the barrier, then the off-node ranks never get through the barrier. Ranks on the same node get through the barrier correctly. In my tests this happens when I run on 8+ compute nodes. Here is the code:

```
#include <mpi.h>
#include <stdio.h>
#include <unistd.h>

int main(int argc, char **argv)
{
  int rank = -1;

  MPI_Init(&argc, &argv);

  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  if (rank == 0) {
    fprintf(stderr, "rank %d sleep\n", rank);
    sleep(1);
  }

  fprintf(stderr, "rank %d >>> Barrier\n", rank);
  MPI_Barrier(MPI_COMM_WORLD);
  fprintf(stderr, "rank %d <<< Barrier 1 DONE\n", rank);

  sleep(100);

  MPI_Finalize();
}
```
and here is the output for 2 ranks per node, 8 compute nodes, using HCOLL:
```
$ mpirun -mca coll_hcoll_enable 1 ./mpitest
rank 12 >>> Barrier
rank 4 >>> Barrier
rank 14 >>> Barrier
rank 10 >>> Barrier
rank 13 >>> Barrier
rank 5 >>> Barrier
rank 0 sleep
rank 8 >>> Barrier
rank 6 >>> Barrier
rank 1 >>> Barrier
rank 9 >>> Barrier
rank 7 >>> Barrier
rank 2 >>> Barrier
rank 15 >>> Barrier
rank 11 >>> Barrier
rank 3 >>> Barrier
rank 0 >>> Barrier
rank 0 <<< Barrier 1 DONE
rank 1 <<< Barrier 1 DONE
```
Attaching to one of the higher ranks with gdb, this is the stack:
```
(gdb) bt
#0  0x00002ac85d482ad0 in mlx5_poll_cq_v1 () from /usr/lib64/libibverbs/libmlx5-rdmav17.so
#1  0x00002ac853893f0c in ibv_poll_cq (wc=0x7fff562500a8, num_entries=<optimized out>, cq=<optimized out>) at /usr/include/infiniband/verbs.h:1245
#2  uct_ib_poll_cq (wcs=0x7fff562500a8, count=<synthetic pointer>, cq=<optimized out>)
    at /scrap/jenkins/workspace/hpc-power-pack/label/r-vmb-rhel7-u6-x86-64-MOFED-CHECKER/hpcx_root/src/hpcx-v2.4.0-gcc-OFED-3.12-1-redhat7.6-x86_64/ucx-v1.6.x/src/uct/ib/base/ib_device.h:280
#3  uct_rc_verbs_iface_poll_rx_common (iface=0x150e880) at rc/verbs/rc_verbs_impl.h:116
#4  uct_rc_verbs_iface_progress (arg=0x150e880) at rc/verbs/rc_verbs_iface.c:131
#5  0x00002ac852d19b8a in ucs_callbackq_dispatch (cbq=<optimized out>)
    at /scrap/jenkins/workspace/hpc-power-pack/label/r-vmb-rhel7-u6-x86-64-MOFED-CHECKER/hpcx_root/src/hpcx-v2.4.0-gcc-OFED-3.12-1-redhat7.6-x86_64/ucx-v1.6.x/src/ucs/datastruct/callbackq.h:211
#6  uct_worker_progress (worker=<optimized out>)
    at /scrap/jenkins/workspace/hpc-power-pack/label/r-vmb-rhel7-u6-x86-64-MOFED-CHECKER/hpcx_root/src/hpcx-v2.4.0-gcc-OFED-3.12-1-redhat7.6-x86_64/ucx-v1.6.x/src/uct/api/uct.h:1806
#7  ucp_worker_progress (worker=0x2ac860945010) at core/ucp_worker.c:1632
#8  0x00002ac860f95be7 in hmca_bcol_ucx_p2p_progress_fast () from /cluster/software/hpcx/2.4/hcoll/lib/hcoll/hmca_bcol_ucx_p2p.so
#9  0x00002ac860f9cf78 in mcast_p2p_recv () from /cluster/software/hpcx/2.4/hcoll/lib/hcoll/hmca_bcol_ucx_p2p.so
#10 0x00002ac861658152 in do_bcast () from /cluster/software/hpcx/2.4/hcoll/lib/hcoll/hmca_mcast_vmc.so
#11 0x00002ac861658d64 in vmc_bcast_multiroot () from /cluster/software/hpcx/2.4/hcoll/lib/hcoll/hmca_mcast_vmc.so
#12 0x00002ac861651125 in hmca_mcast_vmc_bcast_multiroot () from /cluster/software/hpcx/2.4/hcoll/lib/hcoll/hmca_mcast_vmc.so
#13 0x00002ac860f90935 in hmca_bcol_ucx_p2p_bcast_mcast_multiroot () from /cluster/software/hpcx/2.4/hcoll/lib/hcoll/hmca_bcol_ucx_p2p.so
#14 0x00002ac860f90ff7 in bcol_ucx_p2p_ff_barrier () from /cluster/software/hpcx/2.4/hcoll/lib/hcoll/hmca_bcol_ucx_p2p.so
#15 0x00002ac85e6fe48c in hmca_coll_ml_barrier_launch () from /cluster/software/hpcx/2.4/hcoll/lib/libhcoll.so.1
#16 0x00002ac85e6ff191 in hmca_coll_ml_barrier_intra () from /cluster/software/hpcx/2.4/hcoll/lib/libhcoll.so.1
#17 0x00002ac853f821fa in mca_coll_hcoll_barrier (comm=0x402080 <ompi_mpi_comm_world>, module=0x1408010) at coll_hcoll_ops.c:29
#18 0x00002ac84e6beff8 in PMPI_Barrier () at pbarrier.c:74
#19 0x0000000000400994 in main ()
```
When I turn off HCOLL, things work correctly:
```
$ mpirun -mca coll_hcoll_enable 0 ./mpitest
rank 9 >>> Barrier
rank 0 sleep
rank 1 >>> Barrier
rank 3 >>> Barrier
rank 6 >>> Barrier
rank 7 >>> Barrier
rank 15 >>> Barrier
rank 14 >>> Barrier
rank 2 >>> Barrier
rank 8 >>> Barrier
rank 11 >>> Barrier
rank 10 >>> Barrier
rank 12 >>> Barrier
rank 13 >>> Barrier
rank 4 >>> Barrier
rank 5 >>> Barrier
rank 0 >>> Barrier
rank 0 <<< Barrier 1 DONE
rank 1 <<< Barrier 1 DONE
rank 8 <<< Barrier 1 DONE
rank 9 <<< Barrier 1 DONE
rank 4 <<< Barrier 1 DONE
rank 5 <<< Barrier 1 DONE
rank 12 <<< Barrier 1 DONE
rank 2 <<< Barrier 1 DONE
rank 13 <<< Barrier 1 DONE
rank 3 <<< Barrier 1 DONE
rank 10 <<< Barrier 1 DONE
rank 11 <<< Barrier 1 DONE
rank 6 <<< Barrier 1 DONE
rank 14 <<< Barrier 1 DONE
rank 7 <<< Barrier 1 DONE
rank 15 <<< Barrier 1 DONE
```
I played around and it seems that if I put a second barrier after the first one, then all ranks do get through both barriers. So there are clearly some communication issues here.

**Comment by yosefe** [comment]
@jladd-mlnx FYI

**Comment by jladd-mlnx** [comment]
@vspetrov can you take a look, please. Seems stuck in multiroot VMC fan-out.

**Comment by angainor** [comment]
@jladd-mlnx Is there any info about this? Can you guys reproduce it, or is it something about my system? It kind of breaks the idea of a barrier..

**Comment by angainor** [comment]
@jladd-mlnx Thanks. This particular case was the IOR benchmark, which does some sleeps on the master rank sometimes. But in general, I understand that this will also happen when there is work imbalance between the ranks, and e.g., one of them took longer to compute? Or in some code initialization scenarios, when a single rank does dome work, while the rest waits on the barrier? Could be a problem..

**Comment by jladd-mlnx** [comment]
@angainor we added an optimization to our reliability protocol that enables your use case to progress as one might expect. It will be available in our end of September HPC-X v2.5 release.

**Comment by angainor** [comment]
@jladd-mlnx Thanks! 

I see another barrier issue. The following code segfaults and / or hangs when executed with more than 1 OpenMP thread (`mpirun -mca coll_hcoll_enable 1 -x OMP_NUM_THREADS=8 -map-by core -bind-to numa ./mpitest`)
```
#include <stdio.h>                                                                                                                      
#include <mpi.h>                                                                                                                        
#include <omp.h>                                                                                                                        
                                                                                                                                        
int main(int argc, char *argv[])                                                                                                        
{                                                                                                                                       
    int rank, size, threads;                                                                                                            
                                                                                                                                        
    MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &threads);                                                                       
    fprintf(stderr, "thread mode %d %d\n", MPI_THREAD_MULTIPLE, threads);                                                               
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);                                                                                               
    MPI_Comm_size(MPI_COMM_WORLD, &size);                                                                                               
                                                                                                                                        
#pragma omp parallel                                                                                                                    
    {                                                                                                                                   
        int thrid, nthr;                                                                                                                
                                                                                                                                        
        thrid = omp_get_thread_num();                                                                                                   
        nthr = omp_get_num_threads();                                                                                                   
        fprintf(stderr, "rank %d thrid %d started\n", rank, thrid);                                                                     
                                                                                                                                        
#pragma omp barrier                                                                                                                     
        // #pragma omp critical                                                                                                         
        MPI_Barrier(MPI_COMM_WORLD);                                                                                                    
    }                                                                                                                                   
                                                                                                                                        
    MPI_Finalize();                                                                                                                     
}                                                                                                                                       
```
I asked for `MPI_THREAD_MULTIPLE` support, but the barrier does not work unless I turn on the critical region. I get segfaults, or the code hangs. This is the stack from the core dump:
```
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `./mpitest 1000 102400'.
Program terminated with signal 11, Segmentation fault.
#0  0x00002ab3dbf9f3f5 in opal_atomic_fetch_add_32 (delta=-1, addr=0x1) at ../../../../opal/include/opal/sys/gcc_builtin/atomic.h:118
118	    return __atomic_fetch_add (addr, delta, __ATOMIC_RELAXED);
Missing separate debuginfos, use: debuginfo-install bzip2-libs-1.0.6-13.el7.x86_64 elfutils-libelf-0.170-4.el7.x86_64 elfutils-libs-0.170-4.el7.x86_64 glibc-2.17-222.el7.x86_64 hwloc-libs-1.11.10-1.el7.centos.x86_64 libattr-2.4.46-13.el7.x86_64 libcap-2.22-9.el7.x86_64 libevent-2.0.21-4.el7.x86_64 libibcm-41mlnx1-OFED.4.1.0.1.0.44207.x86_64 libibmad-5.0.0.MLNX20180328.afe7c88-0.1.44207.x86_64 libibumad-43.1.1.MLNX20180612.87b4d9b-0.1.44207.x86_64 libibverbs-41mlnx1-OFED.4.4.2.0.1.44207.x86_64 libmlx5-41mlnx1-OFED.4.4.2.0.1.44207.x86_64 libnl3-3.2.28-4.el7.x86_64 librdmacm-41mlnx1-OFED.4.2.0.1.3.44207.x86_64 libtool-ltdl-2.4.2-22.el7_3.x86_64 libxml2-2.9.1-6.el7_2.3.x86_64 numactl-libs-2.0.9-7.el7.x86_64 sssd-client-1.16.0-19.el7_5.8.x86_64 systemd-libs-219-57.el7_5.3.x86_64 xz-libs-5.2.2-1.el7.x86_64 zlib-1.2.7-17.el7.x86_64
(gdb) bt
#0  0x00002ab3dbf9f3f5 in opal_atomic_fetch_add_32 (delta=-1, addr=0x1) at ../../../../opal/include/opal/sys/gcc_builtin/atomic.h:118
#1  opal_atomic_add_fetch_32 (value=-1, addr=0x1) at ../../../../opal/include/opal/sys/atomic_impl.h:384
#2  opal_thread_add_fetch_32 (delta=-1, addr=0x1) at ../../../../opal/threads/thread_usage.h:148
#3  wait_sync_update (updates=1, status=0, sync=0x1) at ../../../../opal/threads/wait_sync.h:112
#4  ompi_request_complete (with_signal=true, request=0xf99728) at ../../../../ompi/request/request.h:451
#5  coll_handle_complete (handle=0xf99728) at coll_hcoll_rte.c:384
#6  0x00002ab3e61a7180 in hmca_coll_ml_barrier_intra () from /cluster/software/hpcx/2.4/hcoll/lib/libhcoll.so.1
#7  0x00002ab3dbfa022a in mca_coll_hcoll_barrier (comm=0x4020c0 <ompi_mpi_comm_world>, module=0x10a4700) at coll_hcoll_ops.c:29
#8  0x00002ab3d79b2018 in PMPI_Barrier () at pbarrier.c:74
#9  0x00002ab3d7a6987e in gomp_thread_start (xdata=<optimized out>) at ../../../libgomp/team.c:120
#10 0x00002ab3d7b4fe25 in start_thread () from /usr/lib64/libpthread.so.0
#11 0x00002ab3d7e62bad in clone () from /usr/lib64/libc.so.6
```
So it seems that HCOLL is involved. Also, the code works if I use `coll_hcoll_enable 0`.

I didn't want to open a new case since the example also deals with `MPI_Barrier`, but let me know if you prefer that I submit a new issue.

**Comment by devreal** [comment]
@angainor The example you provided is not valid in MPI: two threads may not concurrently enter a collective operation (like the `MPI_Barrier`) on the same communicator. You may try using distinct communicators for each thread though. Using the critical region prevents threads from entering the barrier at the same time, so your test becomes a valid MPI application again.

**Comment by angainor** [comment]
@devreal Ups, thanks for this correction. I was obviously fooled by the following statement from the MPI standard:
> MPI_THREAD_MULTIPLE Multiple threads may call MPI, with no restrictions.

The restriction you mention comes from Section 5.13?

> Finally, in multithreaded implementations, one can have more than one, concurrently
executing, collective communication call at a process. In these situations, it is the user’s re-
sponsibility to ensure that the same communicator is not used concurrently by two different
collective communication calls at the same process.

Ehh.. Anyway, thanks again!

**Comment by devreal** [comment]
> `MPI_THREAD_MULTIPLE` Multiple threads may call MPI, with no restrictions.

I guess that is a bit too broad, given that there is also the advice to users in §12.4.1:

> It is the user’s responsibility to prevent races when threads within the same application post conflicting communication calls. The user can make sure that two threads in the same process will not issue conflicting communication calls by using distinct communicators at each thread.

**Comment by gpaulsen** [comment]
Can we close this issue?

As I read it, the hcoll barrier issue has been fixed in newer HCOLL, and the second issue mentioned here is not valid MPI?

**Comment by vspetrov** [comment]
@gpaulsen yes. 1st - fixed an hcoll, 2nd - not a bug.
